In [ ]:
import statsmodels.api as sm    # Loads tools for running statistical tests and regressions
from sklearn import datasets    # Brings in scikit-learn's built-in data utilities
import pandas as pd             # Imports the standard data manipulation library

# RAW github links downloaded and stored here in 2 variables
train_url = 'https://raw.githubusercontent.com/tseringdorjee15/Intro-to-Machine-Learning--2026/refs/heads/main/Boston-Housing/train.csv'
test_url = 'https://raw.githubusercontent.com/tseringdorjee15/Intro-to-Machine-Learning--2026/refs/heads/main/Boston-Housing/test.csv'

# datasets loaded directly into DataFrames
df_train = pd.read_csv(train_url)
df_test = pd.read_csv(test_url)

# Preview data to verify if it is loaded correctly or not
print("--- Training Data Preview ---")
print(df_train.head())                       # print the first 5 rows
print("--- Testing Data Preview ---")
print(df_test.head())                        # print the first 5 rows

--- Training Data Preview ---
   ID     crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0   1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1   2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2   4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
3   5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
4   7  0.08829  12.5   7.87     0  0.524  6.012  66.6  5.5605    5  311   

   ptratio   black  lstat  medv  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     18.7  394.63   2.94  33.4  
3     18.7  396.90   5.33  36.2  
4     15.2  395.60  12.43  22.9  
--- Testing Data Preview ---
   ID     crim    zn  indus  chas    nox     rm    age     dis  rad  tax  \
0   3  0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242   
1   6  0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222   
2   8  0.14455  12.5   7.87     0  0.524  6.172   96.1  5.

In [ ]:
# wE cannot use .DESCR on the data because datasets were loaded from a CSV file into a Pandas DataFrame (df_train) & (df.test).
# The .DESCR attribute only works on built-in datasets that are bundled directly inside the scikit-learn library (like datasets.load_boston()).
# Because our data comes from Kaggle, trying to run df_train.DESCR crashes code with an AttributeError.
# So, we use df_train.info & df_test.info INSTEAD ==> To get data types and column info
df_train.info
df_test.info

<bound method DataFrame.info of       ID     crim    zn  indus  chas    nox     rm    age     dis  rad  tax  \
0      3  0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242   
1      6  0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222   
2      8  0.14455  12.5   7.87     0  0.524  6.172   96.1  5.9505    5  311   
3      9  0.21124  12.5   7.87     0  0.524  5.631  100.0  6.0821    5  311   
4     10  0.17004  12.5   7.87     0  0.524  6.004   85.9  6.5921    5  311   
..   ...      ...   ...    ...   ...    ...    ...    ...     ...  ...  ...   
168  496  0.17899   0.0   9.69     0  0.585  5.670   28.8  2.7986    6  391   
169  497  0.28960   0.0   9.69     0  0.585  5.390   72.9  2.7986    6  391   
170  499  0.23912   0.0   9.69     0  0.585  6.019   65.3  2.4091    6  391   
171  501  0.22438   0.0   9.69     0  0.585  6.027   79.7  2.4982    6  391   
172  505  0.10959   0.0  11.93     0  0.573  6.794   89.3  2.3889    1  273   

     ptratio   black  lstat  
0       17.8  392.83   4.03  
1       18.7  394.12   5.21  
2       15.2  396.90  19.15  
3       15.2  386.63  29.93  
4       15.2  386.71  17.10  
..       ...     ...    ...  
168     19.2  393.29  17.60  
169     19.2  396.90  21.14  
170     19.2  396.90  12.92  
171     19.2  396.90  14.33  
172     21.0  393.45   6.48  

[173 rows x 14 columns]>

In [ ]:
print(df_train.describe())   # To get statistical summaries (mean, min, max, std deviation)
print(df_test.describe())
print(df_train.shape)        # To see the exact shape (number of rows and columns)
print(df_test.shape)

               ID        crim          zn       indus        chas         nox  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean   250.951952    3.360341   10.689189   11.293483    0.060060    0.557144   
std    147.859438    7.352272   22.674762    6.998123    0.237956    0.114955   
min      1.000000    0.006320    0.000000    0.740000    0.000000    0.385000   
25%    123.000000    0.078960    0.000000    5.130000    0.000000    0.453000   
50%    244.000000    0.261690    0.000000    9.900000    0.000000    0.538000   
75%    377.000000    3.678220   12.500000   18.100000    0.000000    0.631000   
max    506.000000   73.534100  100.000000   27.740000    1.000000    0.871000   

               rm         age         dis         rad         tax     ptratio  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean     6.265619   68.226426    3.709934    9.633634  409.279279   18.448048   
std      0.703952   28.1333

In [ ]:
# Split datasets into two separate pieces so that a machine learning model can learn from them.

import numpy as np
import pandas as pd

# Set the features (create a new dataframe named "X_train". It takes original dataset (df_train) and copies everything except the 'medv' (housing price) column.)
X_train = df_train.drop(columns=['medv'])

# Set the target, new dataframe named "Y_train". It isolates just the single column 'medv' into its own separate list
y_train = df_train['medv']

In [ ]:
# SINGLE LINEAR REGRESSION IN "Statsmodels"
# Builds, trains, and tests a Simple Linear Regression model using the statsmodels library to look for a statistical relationship between two variables.

import statsmodels.api as sm

# 1. Using correct DataFrame (df_train) and column name
X = df_train["rm"]           # Sets the number of rooms (rm) as a single predictor feature

y = df_train["medv"]         # Sets the median home value (medv) as the target outcome we want to predict.

# TRAINING THE MODEL - Fit and make the predictions by the model
# sm.OLS: Stands for Ordinary Least Squares. This is the formal math name for a classic linear regression. It calculates a straight "line of best fit"
# through data points..fit(): Tells Python to run the math equations over your training data right now to calculate the weight/slope of the line.
model = sm.OLS(y, X).fit()

# Passes the room counts (X) back into a trained line equation to get the model's generated price guesses.
predictions = model.predict(X)

# Print out the statistics - This is the most important command. It outputs a large, highly detailed statistical report showing our model's accuracy
# (like \(R^{2}\)), coefficients (slopes), and \(p\)-values.
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   medv   R-squared (uncentered):                   0.904
Model:                            OLS   Adj. R-squared (uncentered):              0.903
Method:                 Least Squares   F-statistic:                              3117.
Date:                Wed, 05 Aug 2026   Prob (F-statistic):                   8.25e-171
Time:                        01:33:35   Log-Likelihood:                         -1148.5
No. Observations:                 333   AIC:                                      2299.
Df Residuals:                     332   BIC:                                      2303.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
rm             3.7004      0.066     55.829      0.000       3.570       3.831
==============================================================================
Omnibus:                       66.008   Durbin-Watson:                   0.544
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              129.272
Skew:                           1.049   Prob(JB):                     8.49e-29
Kurtosis:                       5.216   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
X = sm.add_constant(X)     # Addition of the constant term has a direct effect on the coefficient term.  Without it, a model was passing through the origin,
                           # but now we have a y-intercept at -34.67. Also the slope of the RM predictor is also changed from 3.634 to 9.1021 (coef of RM).

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   medv   R-squared:                       0.476
Model:                            OLS   Adj. R-squared:                  0.474
Method:                 Least Squares   F-statistic:                     300.1
Date:                Wed, 05 Aug 2026   Prob (F-statistic):           2.59e-48
Time:                        01:33:35   Log-Likelihood:                -1102.6
No. Observations:                 333   AIC:                             2209.
Df Residuals:                     331   BIC:                             2217.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -33.5367      3.270    -10.254      0.000     -39.970     -27.103
rm             8.9864      0.519     17.324      0.000       7.966      10.007
==============================================================================
Omnibus:                      108.934   Durbin-Watson:                   0.651
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              657.411
Skew:                           1.213   Prob(JB):                    1.76e-143
Kurtosis:                       9.441   Cond. No.                         58.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
# MULTIPLE LINEAR REGRESSION IN "Statsmodels"

# Now we will fit a regression model with more than one variable — "lstat" (percentage of lower status of the population) variable along with the "rm" variable.
# The model training (fitting) procedure remains the exact same as previous:

X = df_train[["rm", "lstat"]]
y = df_train["medv"]

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   medv   R-squared (uncentered):                   0.949
Model:                            OLS   Adj. R-squared (uncentered):              0.949
Method:                 Least Squares   F-statistic:                              3084.
Date:                Wed, 05 Aug 2026   Prob (F-statistic):                   1.03e-214
Time:                        01:33:35   Log-Likelihood:                         -1042.5
No. Observations:                 333   AIC:                                      2089.
Df Residuals:                     331   BIC:                                      2097.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
rm             4.9429      0.087     56.802      0.000       4.772       5.114
lstat         -0.6554      0.038    -17.162      0.000      -0.730      -0.580
==============================================================================
Omnibus:                      125.345   Durbin-Watson:                   0.743
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              451.052
Skew:                           1.647   Prob(JB):                     1.14e-98
Kurtosis:                       7.653   Cond. No.                         4.68
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
# From scikit-learn, import the linear_model class where linear regression can be implemented.
from sklearn import linear_model
X = df_train.drop(columns=["medv"])
y = df_train["medv"]

#X = df_train
#y = df_train["medv"]

lm = linear_model.LinearRegression()
model = lm.fit(X,y)

In [ ]:
predictions = lm.predict(X)
print(predictions[0:5])


[30.47720693 25.12405867 29.35189585 28.43212691 22.99129066]


In [ ]:
lm.score(X,y)


0.7355786478533118

In [ ]:
lm.coef_


array([-4.54789253e-03, -5.17062363e-02,  4.93344687e-02,  5.34084254e-02,
        3.78011391e+00, -1.54106687e+01,  3.87910457e+00, -9.51042267e-03,
       -1.60411361e+00,  3.61780090e-01, -1.14966409e-02, -8.48538613e-01,
        1.18853164e-02, -6.01842329e-01])

**RUNNING 2 LINEAR REGRESSION MODELS SIDE BY SIDE:**

In [ ]:
# MODEL 1- (ALL 13 [FEATURES] EXCLUDING 'medv'[target]) - Trains a multiple linear regression model to predict housing prices.

import statsmodels.api as sm

# Select ALL columns except the target 'medv'
X_all = df_train.drop(columns=['medv'])
X_all = sm.add_constant(X_all) # Add intercept
y = df_train['medv']

model_all = sm.OLS(y, X_all).fit()
print(model_all.summary())


                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.736
Model:                            OLS   Adj. R-squared:                  0.724
Method:                 Least Squares   F-statistic:                     63.19
Date:                Wed, 05 Aug 2026   Prob (F-statistic):           5.73e-83
Time:                        01:33:48   Log-Likelihood:                -988.56
No. Observations:                 333   AIC:                             2007.
Df Residuals:                     318   BIC:                             2064.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         33.7412      6.280      5.373      0.0

In [ ]:
# Cell for Model 2 (Subset: Rooms & Status Only) - trains a simplified regression model using a subset of only two features (rm and lstat) for comparison.

# Select only 'rm' and 'lstat' columns
X_sub = df_train[['rm', 'lstat']]
X_sub = sm.add_constant(X_sub) # Add intercept

model_sub = sm.OLS(y, X_sub).fit()
print(model_sub.summary())


                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.634
Model:                            OLS   Adj. R-squared:                  0.632
Method:                 Least Squares   F-statistic:                     286.2
Date:                Wed, 05 Aug 2026   Prob (F-statistic):           8.11e-73
Time:                        01:33:48   Log-Likelihood:                -1042.5
No. Observations:                 333   AIC:                             2091.
Df Residuals:                     330   BIC:                             2102.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.1039      3.921      0.027      0.9

***Model Comparison and Contrast Summary:***

**Model 1 (All 13 Features):**
    - It achieves a higher R-squared score of 0.736.
    - it captures more complex dynamics and explains an additional 10.2% of the variance in housing prices compared to the simplified model.
    
**Model 2 (Subset Features)**
    - It drops 11 variables and focuses strictly on only 2 features which are: room count (rm) and lower-class status (lstat).
    - While its R-squared score drops to 0.634, it remains a highly valuable   model because it is simple, easy to interpret, and highlights the two single most impactful drivers of real estate value without introducing noise or unnecessary complexity.